### ISSUE TITLE
Unable to load shared library "libAvaloniaNative"

### ISSUE THREAD

**Issue by duck-dev** [AvaloniaUI/Avalonia#9009]
**Describe the bug**
My **self-contained** Avalonia application has been built/published using the publishing features of **JetBrains Rider** under **.NET 6.0.401** on **Pop! OS 21.10** (based on Ubuntu, Linux). The executables for Windows and Linux have always worked perfectly fine. The x64 build for **OSX** however has some **fatal errors** upon trying to open it on a **Mac**. It ended up with several errors in the log file with .NET 6.0.3xx that has been installed with Snap on my Linux machine, this build has been published on, until it settled down to only one error when it was published with .NET 6.0.401 (installed manually, **NOT with Snap**, using https://dotnet.microsoft.com/en-us/download/dotnet and the instructions that appear after downloading the x64 .tar.gz):

```
Unhandled exception. System.DllNotFoundException: Unable to load shared library 'libAvaloniaNative' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: dlopen(liblibAvaloniaNative, 0x0001): tried: 'liblibAvaloniaNative' (no such file), '/usr/local/lib/liblibAvaloniaNative' (no such file), '/usr/lib/liblibAvaloniaNative' (no such file), '/path/to/liblibAvaloniaNative' (no such file)
   at Avalonia.Native.AvaloniaNativePlatform.CreateAvaloniaNative()
   at Avalonia.Native.AvaloniaNativePlatform.Initialize(AvaloniaNativePlatformOptions options) in /_/src/Avalonia.Native/AvaloniaNativePlatform.cs:line 57
   at Avalonia.AvaloniaNativePlatformExtensions.<>c__DisplayClass0_0`1.<UseAvaloniaNative>b__0() in /_/src/Avalonia.Native/AvaloniaNativePlatformExtensions.cs:line 14
   at Avalonia.Controls.AppBuilderBase`1.Setup() in /_/src/Avalonia.Controls/AppBuilderBase.cs:line 303
   at Avalonia.Controls.AppBuilderBase`1.SetupWithLifetime(IApplicationLifetime lifetime) in /_/src/Avalonia.Controls/AppBuilderBase.cs:line 179
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode) in /_/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 208
   at Project.Program.Main(String[] args) in path/to/src/Program.cs:line 13      
```

Earlier, the errors were mostly related to **SkiaSharp** or **`libSystem.Native`**. Regarding those earlier issues, I went through the entire internet, tried to include several packages into the project, like "SkiaSharp.NativeAssets.Linux.NoDependencies", "SkiaSharp.NativeAssets.Linux", "SkiaSharp.NativeAssets.macOS" (all combinations). Unfortunately, none of this lead to success. I followed the advice in many GitHub issues related to this, including #5190, #5862, https://github.com/mono/SkiaSharp/issues/1341 and many more - without any results.

**To Reproduce**
Well, there is not directly a way to reproduce it, but publishing an application with JetBrains Rider for the x64 architecture (OSX) and trying to run the application on a Mac (right-click or control-click the executable file => "Open" => Allow anyway for all files Apple doesn't trust) might yield the same results.

**Expected behavior**
The application should open as usual.

**Desktop (please complete the following information):**

- OS: macOS Monterey Version 12.5 (application published on Pop! OS 21.10)
- Version: Avalonia 0.10.18

**Comment by kekekeks** [comment]
Try running `dotnet publish -r osx-x64 -o build` (or  `dotnet publish -r osx-arm64 -o build`)

**Comment by duck-dev** [comment]
@kekekeks Right, am I supposed to try it out on the Apple device now? That doesn't look much different than my usual steps for publishing. Thank you a lot for your answer!

**Comment by duck-dev** [comment]
@kekekeks Unfortunately, that doesn't change anything. Here are some relevant logs and considering the first error below, I want to emphasize again that the application is published as a self-contained application:

```
Failed to load /path/to/build/libhostfxr.dylib, error: dlopen(/path/to/build/libhostfxr.dylib, 0x0001): tried: '/path/to/build/libhostfxr.dylib' (code signature in <CE6ED5A7-8FCE-3A39-969D-6EE85FDD2267> '/path/to/build/libhostfxr.dylib' not valid for use in process: library load disallowed by system policy)

The library libhostfxr.dylib was found, but loading it from /path/to/build/libhostfxr.dylib failed

  - Installing .NET prerequisites might help resolve this problem.
     https://go.microsoft.com/fwlink/?linkid=2063366
```

```
Failed to load /path/to/build/libhostpolicy.dylib, error: dlopen(/path/to/build/libhostpolicy.dylib, 0x0001): tried: '/path/to/build/libhostpolicy.dylib' (code signature in <43021705-EBF2-36C6-A578-C3FE511CC25B> '/path/to/build/libhostpolicy.dylib' not valid for use in process: library load disallowed by system policy)

An error occurred while loading required library libhostpolicy.dylib from [/path/to/build/]
```

```
Failed to load /path/to/build/libcoreclr.dylib, error: dlopen(/path/to/build/libcoreclr.dylib, 0x0001): tried: '/path/to/build/libcoreclr.dylib' (code signature in <7DAE8258-216E-3A11-A895-DB37B7A647D5> '/path/to/build/libcoreclr.dylib' not valid for use in process: library load disallowed by system policy)

zsh: segmentation fault
```

```
Unhandled exception. System.TypeInitializationException: The type initializer for '<Module>' threw an exception.

---> System.TypeInitializationException: The type initializer for 'Avalonia.MicroCom.MicroComRuntime' threw an exception.
---> System.TypeInitializationException: The type initializer for 'Avalonia.MicroCom.MicroComVtblBase' threw an exception.
---> System.DllNotFoundException: Unable to load shared library 'libSystem.Native' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: dlopen(liblibSystem.Native, 0x0001): tried: 'liblibSystem.Native' (no such file), '/usr/local/lib/liblibSystem.Native' (no such file), '/usr/lib/liblibSystem.Native' (no such file), '/path/to/liblibSystem.Native' (no such file)

   at Interop.Sys.Malloc(UIntPtr size)
   at System.Runtime.InteropServices.NativeMemory.Alloc(UIntPtr byteCount)
   at Avalonia.MicroCom.MicroComVtblBase.CreateVTable()
   at Avalonia.MicroCom.MicroComVtblBase..cctor()

   --- End of inner exception stack trace ---

   at Avalonia.MicroCom.MicroComVtblBase.get_Vtable() in /_/src/Avalonia.MicroCom/MicroComVtblBase.cs:line 16
   at Avalonia.MicroCom.MicroComRuntime..cctor() in /_/src/Avalonia.MicroCom/MicroComRuntime.cs:line 20

   --- End of inner exception stack trace ---

   at Avalonia.MicroCom.MicroComRuntime.Register(Type t, Guid guid, Func`3 proxyFactory) in /_/src/Avalonia.MicroCom/MicroComRuntime.cs:line 30
   at Avalonia.Native.Interop.Impl.__MicroComIAvaloniaNativeFactoryProxy.__MicroComModuleInit() in /_/src/Avalonia.Native/Interop.Generated.cs:line 924
   at .cctor()

   --- End of inner exception stack trace ---

   at Avalonia.AppBuilderDesktopExtensions.LoadAvaloniaNative[TAppBuilder](TAppBuilder builder)
   at Avalonia.AppBuilderDesktopExtensions.UsePlatformDetect[TAppBuilder](TAppBuilder builder) in /_/src/Avalonia.Desktop/AppBuilderDesktopExtensions.cs:line 27
   at Program.BuildAvaloniaApp() in /path/to/src/Program.cs:line 18
   at Program.Main(String[] args) in /path/to/src/Program.cs:line 13
zsh: abort
```

**Comment by berkansasmaz** [comment]
I got similar error when I updated our app to use 11.0.0-preview2, FYI. If necessary, I can share the information you want.

**Comment by kekekeks** [comment]
>library load disallowed by system policy

Are you sure that you have a valid `.app` bundle and that said bundle is allowed to run?

**Comment by duck-dev** [comment]
> Are you sure that you have a valid .app bundle and that said bundle is allowed to run?

@kekekeks No, I packed the entire self-contained application into a ZIP-archive, unpacked it and tried to run it on a Mac, while still making sure to allow the start of this application (the overridden safety settings had to be applied to all the separate files too). Is it absolutely necessary to pack it into a valid `.app` bundle? This still doesn't really explain the exceptions though, does it?

**Comment by berkansasmaz** [comment]
This issue seems to have been resolved for me with 11.0.0-preview3, FYI.

**Comment by hhyyrylainen** [comment]
I was able to get mac exports working by renaming `libAvaloniaNative.dylib` to `liblibAvaloniaNative.dylib` and having `com.apple.security.cs.disable-library-validation` in an entitlements file (and of course having a script to sign all the files after publish). I didn't even have to disable ready to run. That is using .NET 6.
